substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/JfrOldObjectSampler.java

```diff
@@ -0,0 +1,172 @@
+ * questions.
+ */
+
+package com.oracle.svm.core.jfr.oldobject;
+
+import com.oracle.svm.core.Uninterruptible;
+import com.oracle.svm.core.heap.Heap;
+import com.oracle.svm.core.jfr.JfrEvent;
+import com.oracle.svm.core.jfr.JfrTicks;
+import com.oracle.svm.core.jfr.SubstrateJVM;
+import com.oracle.svm.core.locks.SpinLock;
+import com.oracle.svm.core.thread.JavaThreads;
+import jdk.jfr.internal.LogLevel;
+import jdk.jfr.internal.LogTag;
+import jdk.jfr.internal.Logger;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+
+import java.lang.ref.WeakReference;
+
+public final class JfrOldObjectSampler {
+    public static final int DEFAULT_SAMPLER_SIZE = 256;
+
+    private int queueSize;
+    private OldObjectArray samples;
+    private OldObjectPriorityQueue queue;
+    private SpinLock lock;
+    private long lastSweep = Long.MAX_VALUE;
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public JfrOldObjectSampler() {
+    }
+
+    public void configure(int oldObjectQueueSize) {
+        this.queueSize = oldObjectQueueSize;
+    }
+
+    public void initialize() {
+        if (Logger.shouldLog(LogTag.JFR, LogLevel.DEBUG)) {
+            Logger.log(LogTag.JFR, LogLevel.DEBUG, "Initialize old object sampler: old-object-queue-size=" + queueSize);
+        }
+        this.samples = new OldObjectArray(queueSize);
+        this.queue = new OldObjectPriorityQueue(this.samples);
+        this.lock = new SpinLock();
+    }
+
+    @Uninterruptible(reason = "Accesses allocation sampler.")
+    public void sample(WeakReference<Object> ref, long allocatedSize, int arrayLength) {
+        final boolean success = lock.tryLock();
+        if (!success) {
+            return;
+        }
+
+        try {
+            if (queue.isFull()) {
+                if (queue.peek().span > allocatedSize) {
+                    // Sample will not fit, try to scavenge
+                    int numDead = scavenge();
+                    if (numDead == 0) {
+                        // Sample will not fit and all objects still in use, return early
+                        return;
+                    }
+                } else {
+                    // Offered element has a higher priority,
+                    // vacate from the lowest priority one and insert the element.
+                    evict();
+                }
+            }
+
+            store(ref, allocatedSize, JfrTicks.elapsedTicks(), arrayLength);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    // Making callees not uninterruptible to deal with WeakReference.get()
+    @Uninterruptible(reason = "Accesses allocation sampler.", calleeMustBe = false)
+    private int scavenge() {
+        int numDead = 0;
+        for (int i = 0; i < samples.getCapacity(); i++) {
+            final OldObject sample = samples.getSample(i);
+            if (sample.reference != null && sample.reference.get() == null) {
```

Comment:
Yeah, small changes indeed :)

I've not pushed all the commits yet because I was trying to get clarification about https://github.com/oracle/graal/pull/6743#discussion_r1223552890, but that might take some days, so I'll probably push updates shortly.